### ISSUE TITLE
[Console] Error calling command from default command

### ISSUE THREAD

**Issue by tobymackenzie** [symfony/symfony#27838]
**Symfony version(s) affected**: 3.4

**Description**  

[Calling a command](https://symfony.com/doc/current/console/calling_commands.html) that has a required argument from another command works fine normally, but when called from a command being run as the default command, an error such as 'Cannot add a required argument after an optional one' will be thrown.

It may have something to do with the way the `$definition` is built in `Application` when `$name` is falsey, then `Command::mergeApplicationDefinition()` is called on `Command::run()`, but I'm not really sure.

[Possibly related issue](https://github.com/symfony/symfony/issues/25632)

**How to reproduce**  
Create app.php

``` php
<?php
use Symfony\Component\Console\Application;
use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Input\ArrayInput;
use Symfony\Component\Console\Input\InputArgument;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Output\OutputInterface;
require(__DIR__ . '/vendor/autoload.php');

class DefaultCommand extends Command{
	static public $defaultName = 'default';
	protected function execute(InputInterface $input, OutputInterface $output){
		$output->writeln('DefaultCommand');
		$command = $this->getApplication()->find('other');
		$command->run(new ArrayInput([
			'command'=> 'other'
			,'arg'=> 'a'
		]), $output);
	}
}
class OtherCommand extends Command{
	static public $defaultName = 'other';
	protected function configure(){
		$this->addArgument('arg', InputArgument::REQUIRED);
	}
	protected function execute(InputInterface $input, OutputInterface $output){
		$output->writeln('OtherCommand ' . $input->getArgument('arg'));
	}
}

echo "----------\nRunning `" . implode(' ', $_SERVER['argv']) . "`\n";
$app = new Application();
$app->add(new DefaultCommand());
$app->add(new OtherCommand());
$app->setDefaultCommand('default');
$app->run();
```
Run in bash-like shell:

``` sh
composer require symfony/console
php app.php default; php app.php
```

output:

```
----------
Running `app.php default`
DefaultCommand
OtherCommand a
----------
Running `app.php`
DefaultCommand


  Cannot add a required argument after an optional one.


default
```

**Comment by asko-ohmann** [comment]
Hello! 

I tested and debugged this issue and I found the cause is at 

Usually a command definition has argument "command" as REQUIRED but here (when using the default command) the argument "command" is changed to OPTIONAL. https://github.com/symfony/symfony/blob/bf4d011b3d8723657d9c7fdd5cf6d24d74446fef/src/Symfony/Component/Console/Application.php#L212-L221

When combining definitions Application definition is added first. Since the definition now has an OPTIONAL "command" argument no more REQUIRED arguments can be added.
https://github.com/symfony/symfony/blob/bf4d011b3d8723657d9c7fdd5cf6d24d74446fef/src/Symfony/Component/Console/Command/Command.php#L304-L307

Digging through the PR-s it seems this change was introduced in #22244 to fix another issue where default command could not have options.
